### ISSUE TITLE
[Redshift] Merge mode + Parser none Identity column issue

### ISSUE THREAD

**Issue by Olaktal** [embulk/embulk#486]
Hello,

I'm using a redshift db this is my usecase:
I have multiple platforms to store ("France","Spain", ..)

create table platform(
id identity (1,1) int,
original_id varchar(2),
label varchar(50),
active int)

And I have the following data in input:
(original_id,label,active)
('FR','France',1)
('SP','Spain',1)

Original_id is unique and this is my merge key but for performance issues I want to have a integer as primary key so this is why I added an "id" column in my table "platform".

My input config is like:
  parser:
    type: none
    columns:
    - {name: original_id, type: string}
    - {name: label, type: string}
    - {name: active, type: long}

--> It should allow me to insert data into a table with an Identity column

My output config is like:
mode: merge
merge_keys: ["original_id"]

--> This should allow me to merge data if the original_id is already in the platform table, if not, insert it.

But I encountered the following message:

"Error: java.lang.RuntimeException: org.postgresql.util.PSQLException: ERROR: cannot set an identity column to a value"

Do you have any idea on how to solve this problem ?
I tried to add in the output part options like  "incremental: true" and "incremental_columns: [id] " but it didn't worked.
If it is not clear, do not hesitate to ask me.

Best regards,
Vincent

**Comment by hiroyuki-sato** [comment]
Do you use parser type ``none``?.
If so confirm input data with ``embulk preview``

Maybe your input data is the following.
It has not `original_id` column. 

Use parser ``csv`` instead.

```
+--------------------------+
|           payload:string |
+--------------------------+
| original_id,label,active |
|              FR,France,1 |
|               SP,Spain,1 |
|               JP,Japan,1 |
+--------------------------+
```

I succeed to merge data with the following configuration. 
(Note: I use PostgreSQL)

```sql
create table platform(
  id serial not null,
  original_id varchar(2),
  label varchar(50),
  active int
);
ALTER TABLE platform ADD PRIMARY KEY(id);
```

```yaml
in:
  type: file
  path_prefix: ./sample.csv
  parser:
    type: csv
    charset: UTF-8
    newline: CRLF
    delimiter: ','
    quote: '"'
    escape: '"'
    trim_if_not_quoted: false
    skip_header_lines: 1
    allow_extra_columns: false
    allow_optional_columns: false
    columns:
    - {name: original_id, type: string}
    - {name: label, type: string}
    - {name: active, type: long}
out:
  type: postgresql
  user: user
  table: platform
  database: embulk_test
  host: 127.0.0.1
  mode: merge
  merge_keys: ["original_id"]
```

```csv
FR,France,1
SP,Spain,1
JP,Japan,1
```

**Comment by Olaktal** [comment]
Hi @hiroyuki-sato !

The input was mysql but the problem still occur with csv as input.
Embulk preview:

>+--------------------+--------------+-------------+
| original_id:string      | label:string    | active:long  |
+--------------------+--------------+-------------+
|                 ES           |      Espagne   |           1        |
|                 FR           |       France     |           1        |
|                 IT            |       Italie        |           1        |
+--------------------+--------------+-------------+

And this is my YML file

> in:
  type: file
  path_prefix: ./platform.csv
  parser:
    type: csv
    charset: UTF-8
    newline: CRLF
    delimiter: ','
    quote: '"'
    escape: '"'
    trim_if_not_quoted: false
    skip_header_lines: 0
    allow_extra_columns: false
    allow_optional_columns: false
    columns:
    - {name: original_id, type: string}
    - {name: label, type: string}
    - {name: active, type: long}
out:
  type: redshift
  host: *******
  user: *******
  password: *******
  database: *******
  schema: *******
  table: *******
  access_key_id: *******
  secret_access_key: *******
  iam_user_name: *******
  s3_bucket: *******
  s3_key_prefix: *******
  mode: merge
  merge_keys: ["original_id"]

Still have the following message: "ERROR: cannot set an identity column to a value"
Do you think it is a plugin (embulk-output-redshift) malfunction ?

**Comment by hiroyuki-sato** [comment]
I'll ask redshift user about this issue. 
Please wait.

**Comment by Olaktal** [comment]
Thanks @hiroyuki-sato . Note that if I change the mode from merge to insert, it won't crash but obviously all rows are inserted.

**Comment by hiroyuki-sato** [comment]
Could you paste the output of `embulk run -l debug conf.yml` execution for future debug?

I want to see SQL part.

I'll appreciate if you paste log like the below.
(change single "%" character to "`")

%%% 
log paste here
%%%

**Comment by Olaktal** [comment]
I putted the data into a redshift table in the same DB before trying to import it but the result is the same --> 

```
2016-10-07 12:43:58.261 +0000 [INFO] (0001:transaction): Loaded plugin embulk-input-redshift (0.7.3)
2016-10-07 12:43:58.350 +0000 [INFO] (0001:transaction): Loaded plugin embulk-output-redshift (0.6.3)
2016-10-07 12:43:58.490 +0000 [INFO] (0001:transaction): SQL: SET search_path TO ***
2016-10-07 12:43:58.619 +0000 [INFO] (0001:transaction): Using local thread executor with max_threads=4 / output tasks 2 = input tasks 1 * 2
2016-10-07 12:43:58.672 +0000 [INFO] (0001:transaction): Connecting to **** options {user=***, tcpKeepAlive=true, loginTimeout=300, socketTimeout=1800}
2016-10-07 12:43:58.732 +0000 [INFO] (0001:transaction): SQL: SET search_path TO ***
2016-10-07 12:43:58.737 +0000 [INFO] (0001:transaction): > 0.00 seconds
2016-10-07 12:43:58.737 +0000 [INFO] (0001:transaction): Using merge mode
2016-10-07 12:43:58.814 +0000 [INFO] (0001:transaction): SQL: DROP TABLE IF EXISTS "dim_platfor_03b20b80_bl_tmp000"
2016-10-07 12:43:58.818 +0000 [INFO] (0001:transaction): > 0.00 seconds
2016-10-07 12:43:58.914 +0000 [INFO] (0001:transaction): SQL: CREATE TABLE IF NOT EXISTS "dim_platfor_03b20b80_bl_tmp000" ("id" INT4, "original_id" VARCHAR(2), "label" VARCHAR(50), "active" INT4)
2016-10-07 12:43:58.923 +0000 [INFO] (0001:transaction): > 0.01 seconds
2016-10-07 12:43:59.068 +0000 [INFO] (0001:transaction): SQL: DROP TABLE IF EXISTS "dim_platfor_03b20b80_bl_tmp001"
2016-10-07 12:43:59.071 +0000 [INFO] (0001:transaction): > 0.00 seconds
2016-10-07 12:43:59.169 +0000 [INFO] (0001:transaction): SQL: CREATE TABLE IF NOT EXISTS "dim_platfor_03b20b80_bl_tmp001" ("id" INT4, "original_id" VARCHAR(2), "label" VARCHAR(50), "active" INT4)
2016-10-07 12:43:59.182 +0000 [INFO] (0001:transaction): > 0.01 seconds
2016-10-07 12:43:59.302 +0000 [INFO] (0001:transaction): Using merge keys: [original_id]
2016-10-07 12:43:59.319 +0000 [INFO] (0001:transaction): {done:  0 / 1, running: 0}
2016-10-07 12:43:59.362 +0000 [INFO] (0017:task-0000): Connecting to **** options {user=***, tcpKeepAlive=true, loginTimeout=300, socketTimeout=1800}
2016-10-07 12:43:59.716 +0000 [INFO] (0017:task-0000): SQL: SET search_path TO ***
2016-10-07 12:43:59.719 +0000 [INFO] (0017:task-0000): > 0.00 seconds
2016-10-07 12:43:59.719 +0000 [INFO] (0017:task-0000): Copy SQL: COPY "dim_platfor_03b20b80_bl_tmp000" ("original_id", "label", "active") ? GZIP DELIMITER '\t' NULL '\\N' ESCAPE TRUNCATECOLUMNS ACCEPTINVCHARS STATUPDATE OFF COMPUPDATE OFF
2016-10-07 12:43:59.722 +0000 [INFO] (0017:task-0000): Connecting to **** options {user=***, tcpKeepAlive=true, loginTimeout=300, socketTimeout=1800}
2016-10-07 12:43:59.743 +0000 [INFO] (0017:task-0000): SQL: SET search_path TO "***"
2016-10-07 12:43:59.745 +0000 [INFO] (0017:task-0000): > 0.00 seconds
2016-10-07 12:43:59.746 +0000 [INFO] (0017:task-0000): Copy SQL: COPY "dim_platfor_03b20b80_bl_tmp001" ("original_id", "label", "active") ? GZIP DELIMITER '\t' NULL '\\N' ESCAPE TRUNCATECOLUMNS ACCEPTINVCHARS STATUPDATE OFF COMPUPDATE OFF
2016-10-07 12:43:59.804 +0000 [INFO] (0017:task-0000): SQL: SET search_path TO "***"
2016-10-07 12:43:59.807 +0000 [INFO] (0017:task-0000): SQL: DECLARE cur NO SCROLL CURSOR FOR SELECT * FROM "tmp"
2016-10-07 12:43:59.814 +0000 [INFO] (0017:task-0000): SQL: FETCH FORWARD 1000 FROM cur
2016-10-07 12:43:59.821 +0000 [INFO] (0017:task-0000): > 0.01 seconds
2016-10-07 12:43:59.823 +0000 [INFO] (0017:task-0000): SQL: FETCH FORWARD 1000 FROM cur
2016-10-07 12:43:59.825 +0000 [INFO] (0017:task-0000): > 0.00 seconds
2016-10-07 12:43:59.840 +0000 [INFO] (pool-3-thread-1): Uploading file id temp/redshift/a0923635-8fa8-4d7f-a121-382dae0b06c7 to S3 (43 bytes 2 rows)
2016-10-07 12:44:00.269 +0000 [INFO] (pool-3-thread-1): Uploaded file temp/redshift/a0923635-8fa8-4d7f-a121-382dae0b06c7 (0.43 seconds)
2016-10-07 12:44:00.307 +0000 [INFO] (pool-3-thread-2): SQL: SET search_path TO "***"
2016-10-07 12:44:00.310 +0000 [INFO] (pool-3-thread-2): > 0.00 seconds
2016-10-07 12:44:00.310 +0000 [INFO] (pool-3-thread-2): Running COPY from file temp/redshift/a0923635-8fa8-4d7f-a121-382dae0b06c7
2016-10-07 12:44:01.091 +0000 [INFO] (pool-3-thread-2): Loaded file temp/redshift/a0923635-8fa8-4d7f-a121-382dae0b06c7 (0.30 seconds for COPY)
2016-10-07 12:44:01.108 +0000 [INFO] (0017:task-0000): Loaded 1 files.
2016-10-07 12:44:01.108 +0000 [INFO] (pool-4-thread-1): Uploading file id temp/redshift/2cdd8b36-b00f-40cc-89e0-41262dbd00ee to S3 (20 bytes 0 rows)
2016-10-07 12:44:01.177 +0000 [INFO] (pool-4-thread-1): Uploaded file temp/redshift/2cdd8b36-b00f-40cc-89e0-41262dbd00ee (0.07 seconds)
2016-10-07 12:44:01.186 +0000 [INFO] (pool-4-thread-2): SQL: SET search_path TO "dwh"
2016-10-07 12:44:01.189 +0000 [INFO] (pool-4-thread-2): > 0.00 seconds
2016-10-07 12:44:01.189 +0000 [INFO] (pool-4-thread-2): Running COPY from file temp/redshift/2cdd8b36-b00f-40cc-89e0-41262dbd00ee
2016-10-07 12:44:01.803 +0000 [INFO] (pool-4-thread-2): Loaded file temp/redshift/2cdd8b36-b00f-40cc-89e0-41262dbd00ee (0.20 seconds for COPY)
2016-10-07 12:44:01.822 +0000 [INFO] (0017:task-0000): Loaded 1 files.
2016-10-07 12:44:01.825 +0000 [INFO] (0001:transaction): {done:  1 / 1, running: 0}
2016-10-07 12:44:01.826 +0000 [INFO] (0001:transaction): Connecting to **** options {user=***, tcpKeepAlive=true, loginTimeout=300, socketTimeout=28800}
2016-10-07 12:44:01.833 +0000 [INFO] (0001:transaction): SQL: SET search_path TO "**"
2016-10-07 12:44:01.836 +0000 [INFO] (0001:transaction): > 0.00 seconds
2016-10-07 12:44:01.837 +0000 [INFO] (0001:transaction): SQL: BEGIN TRANSACTION;DELETE FROM "dim_platform" USING (SELECT * FROM "dim_platfor_03b20b80_bl_tmp000" UNION ALL SELECT * FROM "dim_platfor_03b20b80_bl_tmp001") S WHERE (S."original_id" = "dim_platform"."original_id");INSERT INTO "dim_platform" (SELECT "original_id", "label", "active" FROM "dim_platfor_03b20b80_bl_tmp000" UNION ALL SELECT "original_id", "label", "active" FROM "dim_platfor_03b20b80_bl_tmp001");END TRANSACTION;
2016-10-07 12:44:01.963 +0000 [ERROR] (0001:transaction): Operation failed (0:0A000)
2016-10-07 12:44:01.971 +0000 [INFO] (0001:cleanup): Connecting to * options {user=***, *** tcpKeepAlive=true, loginTimeout=300, socketTimeout=1800}
2016-10-07 12:44:01.979 +0000 [INFO] (0001:cleanup): SQL: SET search_path TO "***"
2016-10-07 12:44:01.981 +0000 [INFO] (0001:cleanup): > 0.00 seconds
2016-10-07 12:44:01.981 +0000 [INFO] (0001:cleanup): SQL: DROP TABLE IF EXISTS "dim_platfor_03b20b80_bl_tmp000"
2016-10-07 12:44:02.095 +0000 [INFO] (0001:cleanup): > 0.11 seconds
2016-10-07 12:44:02.096 +0000 [INFO] (0001:cleanup): SQL: DROP TABLE IF EXISTS "dim_platfor_03b20b80_bl_tmp001"
2016-10-07 12:44:02.213 +0000 [INFO] (0001:cleanup): > 0.12 seconds
org.embulk.exec.PartialExecutionException: java.lang.RuntimeException: org.postgresql.util.PSQLException: ERROR: cannot set an identity column to a value
        at org.embulk.exec.BulkLoader$LoaderState.buildPartialExecuteException(org/embulk/exec/BulkLoader.java:363)
        at org.embulk.exec.BulkLoader.doRun(org/embulk/exec/BulkLoader.java:572)
        at org.embulk.exec.BulkLoader.access$000(org/embulk/exec/BulkLoader.java:33)
        at org.embulk.exec.BulkLoader$1.run(org/embulk/exec/BulkLoader.java:374)
        at org.embulk.exec.BulkLoader$1.run(org/embulk/exec/BulkLoader.java:370)
        at org.embulk.spi.Exec.doWith(org/embulk/spi/Exec.java:25)
        at org.embulk.exec.BulkLoader.run(org/embulk/exec/BulkLoader.java:370)
        at org.embulk.EmbulkEmbed.run(org/embulk/EmbulkEmbed.java:180)
        at java.lang.reflect.Method.invoke(java/lang/reflect/Method.java:606)
        at org.jruby.javasupport.JavaMethod.invokeDirectWithExceptionHandling(org/jruby/javasupport/JavaMethod.java:453)
        at org.jruby.javasupport.JavaMethod.invokeDirect(org/jruby/javasupport/JavaMethod.java:314)
        at RUBY.run(/home/ubuntu/.embulk/bin/embulk!/embulk/runner.rb:84)
        at RUBY.run(/home/ubuntu/.embulk/bin/embulk!/embulk/command/embulk_run.rb:307)
        at RUBY.<top>(/home/ubuntu/.embulk/bin/embulk!/embulk/command/embulk_main.rb:2)
        at org.jruby.Ruby.runInterpreter(org/jruby/Ruby.java:849)
        at org.jruby.Ruby.loadFile(org/jruby/Ruby.java:2986)
        at org.jruby.RubyKernel.requireCommon(org/jruby/RubyKernel.java:951)
        at org.jruby.RubyKernel.require(org/jruby/RubyKernel.java:944)
        at RUBY.(root)(uri:classloader:/META-INF/jruby.home/lib/ruby/stdlib/rubygems/core_ext/kernel_require.rb:1)
        at home.ubuntu.$_dot_embulk.bin.embulk.embulk.command.embulk_bundle.<top>(file:/home/ubuntu/.embulk/bin/embulk!/embulk/command/embulk_bundle.rb:51)
        at java.lang.invoke.MethodHandle.invokeWithArguments(java/lang/invoke/MethodHandle.java:599)
        at org.jruby.Ruby.runScript(org/jruby/Ruby.java:833)
        at org.jruby.Ruby.runScript(org/jruby/Ruby.java:825)
        at org.jruby.Ruby.runNormally(org/jruby/Ruby.java:760)
        at org.jruby.Ruby.runFromMain(org/jruby/Ruby.java:579)
        at org.jruby.Main.doRunFromMain(org/jruby/Main.java:425)
        at org.jruby.Main.internalRun(org/jruby/Main.java:313)
        at org.jruby.Main.run(org/jruby/Main.java:242)
        at org.jruby.Main.main(org/jruby/Main.java:204)
        at org.embulk.cli.Main.main(org/embulk/cli/Main.java:23)
Caused by: java.lang.RuntimeException: org.postgresql.util.PSQLException: ERROR: cannot set an identity column to a value
        at org.embulk.output.jdbc.AbstractJdbcOutputPlugin.commit(AbstractJdbcOutputPlugin.java:411)
        at org.embulk.output.jdbc.AbstractJdbcOutputPlugin.transaction(AbstractJdbcOutputPlugin.java:348)
        at org.embulk.exec.BulkLoader$4$1$1.transaction(BulkLoader.java:529)
        at org.embulk.exec.LocalExecutorPlugin.transaction(LocalExecutorPlugin.java:54)
        at org.embulk.exec.BulkLoader$4$1.run(BulkLoader.java:524)
        at org.embulk.spi.util.Filters$RecursiveControl.transaction(Filters.java:96)
        at org.embulk.spi.util.Filters.transaction(Filters.java:49)
        at org.embulk.exec.BulkLoader$4.run(BulkLoader.java:518)
        at org.embulk.input.jdbc.AbstractJdbcInputPlugin.transaction(AbstractJdbcInputPlugin.java:193)
        at org.embulk.exec.BulkLoader.doRun(BulkLoader.java:513)
        at org.embulk.exec.BulkLoader.access$000(BulkLoader.java:33)
        at org.embulk.exec.BulkLoader$1.run(BulkLoader.java:374)
        at org.embulk.exec.BulkLoader$1.run(BulkLoader.java:370)
        at org.embulk.spi.Exec.doWith(Exec.java:25)
        at org.embulk.exec.BulkLoader.run(BulkLoader.java:370)
        at org.embulk.EmbulkEmbed.run(EmbulkEmbed.java:180)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at org.jruby.javasupport.JavaMethod.invokeDirectWithExceptionHandling(JavaMethod.java:453)
        at org.jruby.javasupport.JavaMethod.invokeDirect(JavaMethod.java:314)
        at org.jruby.java.invokers.InstanceMethodInvoker.call(InstanceMethodInvoker.java:45)
        at org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:313)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:163)
        at org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:316)
        at org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:77)
        at org.jruby.internal.runtime.methods.MixedModeIRMethod.INTERPRET_METHOD(MixedModeIRMethod.java:109)
        at org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:95)
        at org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:273)
        at org.jruby.runtime.callsite.CachingCallSite.callBlock(CachingCallSite.java:79)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:83)
        at org.jruby.ir.instructions.CallBase.interpret(CallBase.java:423)
        at org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:348)
        at org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:77)
        at org.jruby.ir.interpreter.InterpreterEngine.interpret(InterpreterEngine.java:86)
        at org.jruby.internal.runtime.methods.MixedModeIRMethod.INTERPRET_METHOD(MixedModeIRMethod.java:179)
        at org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:165)
        at org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:197)
        at org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:313)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:163)
        at org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:316)
        at org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:77)
        at org.jruby.ir.interpreter.Interpreter.INTERPRET_ROOT(Interpreter.java:112)
        at org.jruby.ir.interpreter.Interpreter.execute(Interpreter.java:99)
        at org.jruby.ir.interpreter.Interpreter.execute(Interpreter.java:35)
        at org.jruby.ir.IRTranslator.execute(IRTranslator.java:42)
        at org.jruby.Ruby.runInterpreter(Ruby.java:849)
        at org.jruby.Ruby.loadFile(Ruby.java:2986)
        at org.jruby.runtime.load.LibrarySearcher$ResourceLibrary.load(LibrarySearcher.java:235)
        at org.jruby.runtime.load.LibrarySearcher$FoundLibrary.load(LibrarySearcher.java:34)
        at org.jruby.runtime.load.LoadService.tryLoadingLibraryOrScript(LoadService.java:900)
        at org.jruby.runtime.load.LoadService.smartLoadInternal(LoadService.java:540)
        at org.jruby.runtime.load.LoadService.requireCommon(LoadService.java:427)
        at org.jruby.runtime.load.LoadService.require(LoadService.java:393)
        at org.jruby.RubyKernel.requireCommon(RubyKernel.java:951)
        at org.jruby.RubyKernel.require19(RubyKernel.java:944)
        at org.jruby.RubyKernel$INVOKER$s$1$0$require19.call(RubyKernel$INVOKER$s$1$0$require19.gen)
        at org.jruby.internal.runtime.methods.JavaMethod$JavaMethodOneOrNBlock.call(JavaMethod.java:383)
        at org.jruby.internal.runtime.methods.AliasMethod.call(AliasMethod.java:61)
        at org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:313)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:163)
        at org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:316)
        at org.jruby.ir.interpreter.StartupInterpreterEngine.interpret(StartupInterpreterEngine.java:77)
        at org.jruby.ir.interpreter.InterpreterEngine.interpret(InterpreterEngine.java:86)
        at org.jruby.internal.runtime.methods.MixedModeIRMethod.INTERPRET_METHOD(MixedModeIRMethod.java:179)
        at org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:165)
        at org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:197)
        at org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:313)
        at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:163)
        at home.ubuntu.$_dot_embulk.bin.embulk.embulk.command.embulk_bundle.invokeOther66:require(file:/home/ubuntu/.embulk/bin/embulk!/embulk/command/embulk_bundle.rb)
        at home.ubuntu.$_dot_embulk.bin.embulk.embulk.command.embulk_bundle.RUBY$script(file:/home/ubuntu/.embulk/bin/embulk!/embulk/command/embulk_bundle.rb:51)
        at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:599)
        at org.jruby.ir.Compiler$1.load(Compiler.java:111)
        at org.jruby.Ruby.runScript(Ruby.java:833)
        at org.jruby.Ruby.runScript(Ruby.java:825)
        at org.jruby.Ruby.runNormally(Ruby.java:760)
        at org.jruby.Ruby.runFromMain(Ruby.java:579)
        at org.jruby.Main.doRunFromMain(Main.java:425)
        at org.jruby.Main.internalRun(Main.java:313)
        at org.jruby.Main.run(Main.java:242)
        at org.jruby.Main.main(Main.java:204)
        at org.embulk.cli.Main.main(Main.java:23)
Caused by: org.postgresql.util.PSQLException: ERROR: cannot set an identity column to a value
        at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2182)
        at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:1911)
        at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:173)
        at org.postgresql.jdbc2.AbstractJdbc2Statement.execute(AbstractJdbc2Statement.java:645)
        at org.postgresql.jdbc2.AbstractJdbc2Statement.executeWithFlags(AbstractJdbc2Statement.java:481)
        at org.postgresql.jdbc2.AbstractJdbc2Statement.executeUpdate(AbstractJdbc2Statement.java:409)
        at org.embulk.output.jdbc.JdbcOutputConnection.executeUpdate(JdbcOutputConnection.java:468)
        at org.embulk.output.jdbc.JdbcOutputConnection.collectMerge(JdbcOutputConnection.java:343)
        at org.embulk.output.jdbc.AbstractJdbcOutputPlugin.doCommit(AbstractJdbcOutputPlugin.java:702)
        at org.embulk.output.jdbc.AbstractJdbcOutputPlugin$2.run(AbstractJdbcOutputPlugin.java:404)
        at org.embulk.output.jdbc.AbstractJdbcOutputPlugin$RetryableSQLExecution.call(AbstractJdbcOutputPlugin.java:1070)
        at org.embulk.output.jdbc.AbstractJdbcOutputPlugin$RetryableSQLExecution.call(AbstractJdbcOutputPlugin.java:1058)
        at org.embulk.spi.util.RetryExecutor.run(RetryExecutor.java:100)
        at org.embulk.spi.util.RetryExecutor.runInterruptible(RetryExecutor.java:77)
        at org.embulk.output.jdbc.AbstractJdbcOutputPlugin.withRetry(AbstractJdbcOutputPlugin.java:1042)
        at org.embulk.output.jdbc.AbstractJdbcOutputPlugin.withRetry(AbstractJdbcOutputPlugin.java:1035)
        at org.embulk.output.jdbc.AbstractJdbcOutputPlugin.commit(AbstractJdbcOutputPlugin.java:399)
        ... 82 more

Error: java.lang.RuntimeException: org.postgresql.util.PSQLException: ERROR: cannot set an identity column to a value

```

**Comment by Olaktal** [comment]
Hello @hiroyuki-sato 
Any news about this issue ? If I can do anything  do not hesitate to tell me :)
Best regards,
Vincent

**Comment by Olaktal** [comment]
Hello @hito4t,
I saw you are very active for redshift issues, can you help me on this one ?
Thanks a lot

**Comment by hiroyuki-sato** [comment]
I'm asking @hito4t about this issue at last weekend 😄 


Could you try the following SQL on Redshift with psql? 

## Prepare sample data

```sql
create table platform(
  id identity (1,1) int,,
  original_id varchar(2),
  label varchar(50),
  active int
);
-- I don't know this SQL work on Redshift.
-- ALTER TABLE platform ADD PRIMARY KEY(id);
```

```sql
create table platform_tmp(
  original_id varchar(2),
  label varchar(50),
  active int
);
```

```sql
insert into platform_tmp(original_id,label,active) values('FR','France',1);
insert into platform_tmp(original_id,label,active) values('SP','Spain',1);
```

## Test1

When you execute this SQL, Will you get the error ``cannot set an identity column to a value``

```sql
INSERT INTO platform
  (SELECT original_id,label,active 
     FROM platform_tmp);
```

## Test2

How about this?

```sql
INSERT INTO platform(original_id,label,active)
  (SELECT original_id,label,active 
     FROM platform_tmp);
```
## Test3

If test1 fail and test2 OK, The following SQL might work. 
And Probably we need to change insert statement.

```sql
BEGIN TRANSACTION;
DELETE FROM "platform" 
 USING 
  (SELECT * 
    FROM "platform_tmp" ) S
  WHERE (S."original_id" = "platform"."original_id");
INSERT INTO platform(original_id,label,active)
  (SELECT original_id,label,active 
     FROM platform_tmp);
END TRANSACTION;
```

**Comment by hito4t** [comment]
Hi
embulk-output-redshift merge mode generates SQL as follows.
<pre>
INSERT INTO "platform" (SELECT "original_id", "label", "active" FROM "platform_xxxxxxxx_tmp")
</pre>

But this SQL doesn't work if an identity column exists.
(It seems that Redshift tries to set the `id` column to the value from `platform_xxxxxxxx_tmp.original_id`.)

The following SQL will work.
<pre>
INSERT INTO "platform"("original_id", "label", "active") (SELECT "original_id", "label", "active" FROM "platform_xxxxxxxx_tmp")
</pre>

I'll try to fix this issue.

**Comment by hiroyuki-sato** [comment]
@hito4t 

PostgreSQL work fine. Because It's use the following SQL.

```sql
WITH updated AS (UPDATE "platform" SET "original_id" = S."original_id", "label" = S."label", "active" = S."active" FROM (SELECT "original_id", "label", "active" FROM "platform_763179a7b00_bl_tmp000" UNION ALL SELECT "original_id", "label", "active" FROM "platform_763179a7b00_bl_tmp001" UNION ALL SELECT "original_id", "label", "active" FROM "platform_763179a7b00_bl_tmp002" UNION ALL SELECT "original_id", "label", "active" FROM "platform_763179a7b00_bl_tmp003") S WHERE "platform"."original_id" = S."original_id" RETURNING S."original_id") INSERT INTO "platform" ("original_id", "label", "active") SELECT DISTINCT ON ("original_id") * FROM (SELECT "original_id", "label", "active" FROM "platform_763179a7b00_bl_tmp000" UNION ALL SELECT "original_id", "label", "active" FROM "platform_763179a7b00_bl_tmp001" UNION ALL SELECT "original_id", "label", "active" FROM "platform_763179a7b00_bl_tmp002" UNION ALL SELECT "original_id", "label", "active" FROM "platform_763179a7b00_bl_tmp003") S WHERE NOT EXISTS (SELECT 1 FROM updated WHERE S."original_id" = updated."original_id")
```

**Comment by hito4t** [comment]
Thank you!

I can't reproduce the issue with MySQL because embulk-output-mysql doesn't support `merge_key`.
Anyway, the SQL generated by embulk-output-mysql specifies columns to insert.

embulk-output-oracle and embulk-output-sqlserver doesn't support merge mode yet.

**Comment by hito4t** [comment]
I've created the PR https://github.com/embulk/embulk-output-jdbc/pull/137 .

**Comment by hiroyuki-sato** [comment]
@Olaktal 
Could you try the following command?
Or please wait for next release.

You need Java development kit.

```
git clone https://github.com/embulk/embulk-output-jdbc
cd embulk-output-jdbc
git checkout -b redshift_merge_columns origin/redshift_merge_columns
./gradlew classpath
```

```
embulk run -I /path/to/embulk-output-jdbc/embulk-output-redshift/lib/ conf.yml
```

@hito4t 
Thank you for the working.

**Comment by Olaktal** [comment]
It works great !! Thanks a lot to both of you :)

Do you know when the release will be available ?
I saw you modified the file 'RedshiftOutputConnection.java' so hopefully next time I won't bother you and I'll resolve this kind of issue myself :)

Have a good day

**Comment by hito4t** [comment]
I've released embulk-output-redshift 0.6.5 !
https://github.com/embulk/embulk-output-jdbc/blob/master/ChangeLog

**Comment by Olaktal** [comment]
Thanks a lot for your quickness @hito4t 

Have a good day
